### DISCUSSION TITLE
Roda and Faye websocket

### DISCUSSION THREAD

**Discussion by jeezs** [jeremyevans/roda#246] [General]
When adding the Faye/websocket plugin I have the following error :  Invalid plugin type: String (Roda::RodaError)

It used to work correctly a few month month ago.

Just trying this very simple example thrown this error :
` 
# cat config.ru
require "roda"
require "faye/websocket"

class App < Roda
	plugin "faye/websocket", adapter: :thin, ping: 45
    route do |r|
		"<h1>Hello</h1>"
	end
end

run App.freeze.app
`

Can someone help ?
Do I need to open an issue.
Thanks

**Comment by jeremyevans** [comment]
I don't think this could have ever worked correctly.  The first argument to `plugin` needs to be a symbol or a module.  Roda doesn't ship a `faye/websocket` plugin, so even if the `plugin` call worked , you would have to discuss any issues with the author of that plugin.

**Reply by jeezs** [reply]
Thanks for your answer.
It used to work perfectly it is  only broken since some recents release. 
I try to "symbolise" Faye/websocket but have no luck with it.
Anyway I don't need to use Faye/websocket if you have an alternative websocket solution working with Roda.

**Reply by jeremyevans** [reply]
It's impossible for the code you have posted to have worked in any version of Roda.  If you are looking for a websocket solution for Roda, try https://github.com/socketry/roda-websockets

**Reply by jeezs** [reply]
thanks a lot  I'll use socketry!
As strange as can be I can tell that it was working and even still working on my test server.
I will investigate how this was possible, just for the fun, and will report if your are interested.

**Reply by jeremyevans** [reply]
The test server must be using different code than you posted, or you are misremembering things.  But please, investigate and report what you find.

**Comment by jeezs** [comment]
Here is the full code used on the server : 
`
if RUBY_PLATFORM == "x64-mingw32"
  require "em/pure_ruby"
end
require "sequel"
require "rufus-scheduler"
require "faye/websocket"
require "json"
require "securerandom"
require 'mail'
class String
  def is_json?
	begin
	  !!JSON.parse(self)
	rescue
	  false
	end
  end
end
class App < Roda
  @@channels = {}
  @@user
  eden = Sequel.connect("sqlite://eden_doors.sqlite3")
  unless File.exist?("eden_doors.sqlite3")
	eden.create_table :objects do
	  primary_key :atome_id
	  String :id
	  String :type
	  String :name
	  String :content
	end
  end
  index_content = File.read("public/index.html")
  index_content = index_content.gsub('<script type="text/javascript" src="../cordova.js"></script>', "")
  plugin :static, %w[/css /js /medias]
  plugin "faye/websocket", adapter: :thin, ping: 45

  route do |r|
	if Faye::WebSocket.websocket?(env)
	  ws = Faye::WebSocket.new(env)
	  ws.on :message do |event|
		client_data = event.data
		if client_data.is_json?
		  data = JSON.parse(client_data)
		  case data["type"]
		  when "login"
			user_id = data["username"]
			session_id = SecureRandom.uuid
			message_back = JSON.generate({ type: :response, request_id: data["request_id"], session_id: session_id, log: true })
			ws.send(message_back)
		  when "start_channel"
			channel_id = SecureRandom.uuid
			@@channels[channel_id] = []
			message_back = JSON.generate({ type: :response, request_id: data["request_id"], channel_id: channel_id })
			ws.send(message_back)
		  when "list_channels"
			message_back = JSON.generate({ type: :response, request_id: data["request_id"], channels: @@channels.keys })
			ws.send(message_back)
		  when "connect_channel"
			channel_id = data["channel_id"]
			@@channels[channel_id] << ws
			message_back = JSON.generate({ type: :response, request_id: data["request_id"], connected: true })
			ws.send(message_back)
		  when "push_to_channel"
			channel_id = data["channel_id"]
			message_received = data["message"]
			# fixme: the type depend on the kind if received message
			message_to_push = JSON.generate({ type: :code, content: message_received })
			@@channels[channel_id].each do |ws_found|
			  # we exclude the sender from the recipient
			  unless ws_found == ws
				ws_found.send(message_to_push)
			  end
			end
			message_back = JSON.generate({ type: :response, request_id: data["request_id"], pushed: true })
			ws.send(message_back)
		  when "read"
			file_content = File.read(data["file"])
			hashed_content = { content: file_content }
			hashed_options = { options: data["options"].to_s }
			message_to_push = JSON.generate({ type: :read, atome: data["atome"], target: data["target"], content: hashed_content, options: hashed_options })
			ws.send(message_to_push)
		  when "list"
			files_found = Dir[data["path"] + "/*"]
			hashed_content = { content: files_found }
			hashed_options = { options: data["options"].to_s }
			message_to_push = JSON.generate({ type: :read, target: data["target"], content: hashed_content, options: hashed_options })
			ws.send(message_to_push)
		  when "write"
			File.write(data["file"], data["content"])
			hashed_content = { content: data["content"].to_s }
			hashed_options = { options: data["options"].to_s }
			message_to_push = JSON.generate({ type: :read, target: data["target"], content: hashed_content, options: hashed_options })
			ws.send(message_to_push)
		  when "copy"
			FileUtils.cp data["source"], data["dest"]
		  when "delete"
			File.delete(data["file"])
			hashed_content = { content: data["file"].to_s }
			hashed_options = { options: data["options"].to_s }
			message_to_push = JSON.generate({ type: :read, target: data["target"], content: hashed_content, options: hashed_options })
			ws.send(message_to_push)
		  when "mail"
			if data["from"]
			  sender = data["from"]
			else
			  sender = "contact@atome.one"
			end
			receiver = data["to"]
			mail_subject = data["subject"]
			content = data["content"]
			attachment = data["attachment"]
			attachments = []
			if attachment
			  if attachment.instance_of?(Array)
				attachment.each do |file|
				  filename = File.basename(file)
				  attachments << { file: file, filename: filename }
				end
			  else
				filename = File.basename(attachment)
				attachments << { file: attachment, filename: filename }
			  end
			end
			mail = Mail.new do
			  from sender
			  to receiver
			  subject mail_subject
			  body content
			  attachments.each do |file_to_add|
				file = file_to_add[:file]
				filename = file_to_add[:filename]
				add_file :filename => filename, :content => File.read(file)
			  end
			end
			mail.delivery_method :sendmail
			mail.deliver
		  when "atome"
			message_to_push = JSON.generate({ type: :atome, target: data["target"], atome: data["atome"], content: data["content"] })
			ws.send(message_to_push)
		  when "code"
			message_to_push = JSON.generate({ type: :code, content: data["content"] })
			ws.send(message_to_push)
		  when "command"
			file_content = `#{data["content"]}`
			message_to_push = JSON.generate({ type: :command, target: data["target"], atome: data["atome"], content: file_content })
			ws.send(message_to_push)
		  else
			ws.send("unknown message received")
		  end
		end
	  end
	  ws.rack_response
	end
	r.root do
	  r.redirect "/index"
	end
	r.on "index" do
	  r.is do
		r.get do
		  #sleep 7
		  #box()
		  index_content
		end
	  end
	end
  end
end
`


Here the list of the gem used on the server : 

`
ast (2.4.2)
bundler (2.2.16)
c_lexer (2.6.4.1.1)
concurrent-ruby (1.1.9)
et-orbi (1.2.4)
eventmachine (1.2.7)
execjs (2.8.1, 2.7.0)
faye-websocket (0.11.0)
filewatcher (1.1.1)
fugit (1.5.0, 1.4.5)
image_size (2.1.1, 2.1.0)
jaro_winkler (1.5.4)
mail (2.7.1)
mini_mime (1.1.0)
nio4r (2.5.7)
opal (1.2.0, 1.0.5)
opal-jquery (0.4.5, 0.4.4)
optimist (3.0.1)
parallel (1.20.1)
parser (3.0.2.0, 2.7.2.0, 2.6.4.1)
puma (5.0.2)
raabro (1.4.0)
rack (2.2.3)
rack-unreloader (1.7.0)
rainbow (3.0.0)
rake (13.0.6, 13.0.3)
regexp_parser (2.1.1)
rexml (3.2.5)
roda (2.26.0)
rubocop (1.18.4, 1.7.0, 0.78.0)
rubocop-ast (1.8.0, 1.4.1)
ruby-progressbar (1.11.0)
rufus-scheduler (3.8.0, 3.7.0)
sequel (5.46.0, 5.45.0, 5.43.0, 5.42.0)
sqlite3 (1.4.2)
tzinfo (2.0.4)
uglifier (4.2.0)
unicode-display_width (2.0.0, 1.7.0, 1.6.1)
websocket-driver (0.7.5, 0.7.4, 0.7.3)
websocket-extensions (0.1.5)
`
I will try to reinstall faye-websocket (0.11.0) and roda (2.26.0) on my current code code to see what happen.

**Reply by jeremyevans** [reply]
The Roda version you are using is over 4 years out of date, and no longer supported.  I wouldn't have thought it would work, but looking at the implementation of `plugin` in that version (https://github.com/jeremyevans/roda/blob/2.26.0/lib/roda.rb#L176), it would ignore a string instead of giving an error.  You could just remove the `plugin "faye/websocket", adapter: :thin, ping: 45` as it doesn't do anything.  Maybe the rest of the code will work still work with current Roda.

**Comment by jeezs** [comment]
I 've found:
I've revert Roda gem to  "2.26.0" and it work.
So it may be an issue. 
Tell me what you think.

**Reply by jeezs** [reply]
Just try several things but didn't succeed to make it work with the latest version.
Do you think it's possible for you to a make it work?
Thanks

**Reply by jeremyevans** [reply]
Can you list what error you are getting and what you tried to fix it?

**Reply by jeezs** [reply]
First I just try removing  "adapter: :thin, ping: 45"
like this : 

`
require "roda"
require "faye/websocket"

class App < Roda
plugin  "faye/websocket".
`

So I have the following error : 

`
/Users/jean-ericgodard/.rvm/gems/ruby-3.0.0/gems/roda-3.52.0/lib/roda.rb:282:in `plugin': Invalid plugin type: String (Roda::RodaError)
	from /Users/jean-ericgodard/RubymineProjects/atome/www/app.rb:62:in `<class:App>'
	from /Users/jean-ericgodard/RubymineProjects/atome/www/app.rb:33:in `<top (required)>'
	from /Users/jean-ericgodard/.rvm/gems/ruby-3.0.0/gems/rack-unreloader-1.8.0/lib/rack/unreloader/reloader.rb:310:in `require'
	from /Users/jean-ericgodard/.rvm/gems/ruby-3.0.0/gems/rack-unreloader-1.8.0/lib/rack/unreloader/reloader.rb:310:in `safe_load'
	from /Users/jean-ericgodard/.rvm/gems/ruby-3.0.0/gems/rack-unreloader-1.8.0/lib/rack/unreloader/reloader.rb:201:in `block in require_dependencies'
	from /Users/jean-ericgodard/.rvm/gems/ruby-3.0.0/gems/rack-unreloader-1.8.0/lib/rack/unreloader/reloader.rb:190:in `each'
	from /Users/jean-ericgodard/.rvm/gems/ruby-3.0.0/gems/rack-unreloader-1.8.0/lib/rack/unreloader/reloader.rb:190:in `require_dependencies'
	from /Users/jean-ericgodard/.rvm/gems/ruby-3.0.0/gems/rack-unreloader-1.8.0/lib/rack/unreloader.rb:95:in `require'
`


Then I just try to convert "faye/websocket" to symbol:
(I know the chance to make this work were slim)

`
require "roda"
require "faye/websocket"

class App < Roda

new_socket = "faye/websocket".to_sym
plugin new_socket

`

But it doesn't work too.

**Reply by jeremyevans** [reply]
I'm sorry if my earlier response was not clear.  When I said remove `plugin "faye/websocket", adapter: :thin, ping: 45`, it means remove the entire line.  As I explained, the line does nothing in Roda 2.26.0.

**Reply by jeezs** [reply]
A big big thanks for your precious help.
Everything seems to work, and I can now use the latest version of Roda!
